### ISSUE TITLE
Add support for minimum amount of approvals before merge

### ISSUE THREAD

**Issue by fmvilas** [asyncapi/.github#190]
Have a look at https://github.com/asyncapi/spec/pull/851#issuecomment-1283573288.

**Comment by Amzani** [comment]
I would like to work on this issue.

**Comment by Amzani** [comment]
@fmvilas @derberg You want to require a minimum of 3 approvals for all AsyncAPI repos or you want to only to do it in some repos like `specs` ?

cc @KhudaDad414

**Comment by KhudaDad414** [comment]
@Amzani I think we only need it to be 3 for `specs`.
Besides we get the value from an environment variable so each repo can have their own value :)

**Comment by Amzani** [comment]
Great @KhudaDad414 this environment exists already ? or it's something new we have to create ?

**Comment by octonawish-akcodes** [comment]
Can I try this.

**Comment by derberg** [comment]
> Can I try this.

first we need to figure the solution.

---
the best if that info is stored in a file in a repo, then any change goes through review. 
there is this concept of https://github.com/asyncapi/.github/issues/137 that we need anyway, not just amount of reviewers but for transparency reasons and increasing size of maintainers best if any repo setting is done through a PR

**Comment by derberg** [comment]
yes, we need to have it per repo definitely (like https://github.com/repository-settings/app I guess), so only codeowners of given repo are involved in review, and they can adjust settings transparently however they need for they repository. Of course if there is a case that there are multiple repos that have exactly the same configuration, we can have it in `.github` repo and push out to repos that need it using the same automation we have for workflows or code of conduct. I'm referring to https://github.com/asyncapi/.github/blob/master/.github/workflows/global-replicator.yml

On requirements side, user management would be lovely too 😄 Basically anything that requires going to `Settings` so we can eliminate a need to access `Settings` in general.

> Custom configuration (sonarcloud: true or coveralls: true) will be only possible in [Safe-settings](https://github.com/github/safe-settings). the [repository-settings](https://github.com/repository-settings/app) doesn't support it, However it can be done by developing our custom app on top of [probot](https://github.com/probot/probot) and [octokit-plugin-config](https://github.com/probot/octokit-plugin-config)

For others, like sonar or coveralls, yeah, we definitely will need something custom.

Most important is that we need a strong requirement here to make sure it all works like all else on GitHub Actions. So there are no apps, no deployments and related overhead. Just workflows that support sync of settings whenever they are modified. Maybe solution is to get important code from `repository-setting` and use in our custom action. Or maybe there is some action out there that we could already use or contribute to. 

Afaik with existing mutations in GitHub GraphQL it is not that complicated after all. Updating all stuff with one shot is with one call. So the only complicated part I think is transform of YAML settings to API call.

**Comment by derberg** [comment]
it is definitely not developed yet

default should be 1 as it is now because of how things work, and per repo can be overwritten. 

make sure to first explain how you want to solve it, before you open a PR and hit our review 😃

**Comment by aakankshabhende** [comment]
@derberg I believe that we can have a central config file in this repo for minimum approval, then for individual repo we can include a check if the that repository has its own override config (workflow), if not we will enforce the central workflow. I would love to hear any thoughts or ideas on how we can refine this approach.

**Comment by derberg** [comment]
@aakankshabhende please take into account that in AsyncAPI we use https://github.com/derberg/manage-files-in-multiple-repositories action that allows us to push workflows from `.github` to all the other repositories under `asyncapi` org. So we can have everywhere the same workflow that will be responsible for this approvals verification. No need for any central workflow.

As if it comes to configuration, instead of the config file, just like in some other issue Fran wrote, we can just have a system variable, that GitHub allows to have. So we have https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables on organization level, and workflow would read this variable, and if someone in some repository wants to overwrite this global default, they just overwrite the variable with custom value only in their repository

**Comment by aakankshabhende** [comment]
Got it. Sounds good! Thanks @derberg

**Comment by aakankshabhende** [comment]
I'll start working on this

**Comment by derberg** [comment]
awesome

please feel free to ask for any clarifications on the way - always better to clarify first than get ping-pong in a PR

if I do not answer here for some reason, feel free to DM me in AsyncAPI Slack

good luck!
